### PR TITLE
Add friends and call cancel notifications

### PR DESCRIPTION
**Pull request by hauntii**
This PR adds the following:

- Out of app incoming call notifications via gRPC to the Friends server
- In-app call cancellation notifications
- Standardizes notification processor (hopefully thats what that comment meant)

### CODE REVIEW

File: grpc/send_incoming_call_notification.go
Comment (DaniElectra) [inline grpc/send_incoming_call_notification.go]: We recently added globals for [nintendo notification types](https://github.com/PretendoNetwork/nex-protocols-go/blob/7a12fc5e4b85a1e4adbda6ea50a5fc56f26a55a4/nintendo-notifications/notification_types.go#L39C12-L39C12) in nex-protocols-go to make event types more clear. If possible, please update the modules and use that instead

File: go.mod
Comment (DaniElectra) [inline go.mod]: Not sure how did this happen? This requirement was put in the indirect requirements and another `require ()` field was created

Comment (hauntii) [reply go.mod]: VSCode decided it didn't want to play nice today. Fixed!

File: grpc/send_incoming_call_notification.go
Comment (jonbarrow) [inline grpc/send_incoming_call_notification.go]: Creating a new connection every single time isn't very efficient. I would suggest moving the client somewhere globally accessible, and only establishing the connection once at startup

File: grpc/send_incoming_call_notification.go
Comment (jonbarrow) [inline grpc/send_incoming_call_notification.go]: This is only relevant once your plogger PR is merged, and I'll only write this once despite it being in multiple files

For official logging we prefer to stick with plogger whenever possible, that way all servers log files are consistent and have all information required to look into issues

File: nex/matchmake-extension/update_notification_data.go
Comment (jonbarrow) [inline nex/matchmake-extension/update_notification_data.go]: Only gonna comment this once despite it being elsewhere

Do we know if these types are specific to WiiU Chat? Or are they common to the matchmaking protocol as a whole?

If they're common to the protocol then I'm thinking we should make a struct for them there just like the notifications types

If they're specific to WiiU Chat, then a struct for their types should be made here instead, for clarity's sake so it's clear at a glance what each types value means

**Review by jonbarrow — CHANGES_REQUESTED** [review]
Gonna at least wait until the plogger PR is merged so that the logging here can be moved to plogger

File: nex/matchmake-extension/update_notification_data.go
Comment (DaniElectra) [reply nex/matchmake-extension/update_notification_data.go]: > If they're common to the protocol then I'm thinking we should make a struct for them there just like the notifications types

In case this is common to the protocol (which I think it is), we don't have to make a special struct. Technically our documentation about notification types is *wrong*, because we store the whole number without splitting anything `109000`. However, [other Rendez-Vous servers](https://github.com/zeroKilo/GROBackendWV/wiki/Notifications) document notification types as `type * 1000 + subtype` (for example `109 * 1000 + 0`).

In this function we get the type without the subtype `101`, which is equivalent to the notification type `101000` (this notification looks like an invitation to join a gathering?)

Comment (jonbarrow) [reply nex/matchmake-extension/update_notification_data.go]: I forgot about this! Our documentation and implementations need to be updated then

**Comment by jonbarrow** [comment]
Plogger has been updated so this PR can continue now

File: nex/matchmake-extension/update_notification_data.go
Comment (jonbarrow) [reply nex/matchmake-extension/update_notification_data.go]: `nex-protocols-go` v1.0.36 has been released which changes how notification types are handled. Types are now split into categories and subtypes, the way they are implemented in other Rendezvous servers

This release also includes a helper function `BuildNotificationType`, which takes in a category and a subtype and returns `(category * 1000) + subtype`

Proper usage is now:

```go
notificationType := notifications.BuildNotificationType(notifications.NotificationCategories.RequestJoinGathering, notifications.NotificationSubTypes.RequestJoinGathering.None)
```

Or more broken up:

```go
category := notifications.NotificationCategories.RequestJoinGathering
subtype := notifications.NotificationSubTypes.RequestJoinGathering.None
notificationType := notifications.BuildNotificationType(category, subtype)
```

Please update the notifications handling to use this new release

File: nex/matchmake-extension/update_notification_data.go
Comment (jonbarrow) [inline nex/matchmake-extension/update_notification_data.go]: Type 101 is [defined in `nex-protocols-go`](https://github.com/PretendoNetwork/nex-protocols-go/blob/master/notifications/notification_categories.go#L25), please use the constant

File: nex/matchmake-extension/update_notification_data.go
Comment (jonbarrow) [inline nex/matchmake-extension/update_notification_data.go]: Type 102 is [defined in `nex-protocols-go`](https://github.com/PretendoNetwork/nex-protocols-go/blob/master/notifications/notification_categories.go#L26), please use the constant

File: grpc/send_incoming_call_notification.go
Comment (jonbarrow) [inline grpc/send_incoming_call_notification.go]: This was marked as resolved, but `log` is still being used when Plogger is preferred. Logs from the `log` module will not be tracked in the usual log files we manage. This appears in other files as well, only commenting it once

File: grpc/send_incoming_call_notification.go
Comment (jonbarrow) [inline grpc/send_incoming_call_notification.go]: If this is not a PID, what does 4 represent here? Is this known or are you just replicating a real server response?

File: grpc/send_incoming_call_notification.go
Comment (jonbarrow) [inline grpc/send_incoming_call_notification.go]: Do we know what flags these are? I'm wondering if we should build some kind of flag building system as well for these

File: grpc/send_incoming_call_notification.go
Comment (jonbarrow) [inline grpc/send_incoming_call_notification.go]: Is this related to the 4 value above?

File: nex/matchmake-extension/get_friend_notification_data.go
Comment (jonbarrow) [inline nex/matchmake-extension/get_friend_notification_data.go]: Unnecessary parentheses (I know this was not done by you but it showed in your PR so I'm commenting anyway)

**Review by jonbarrow — CHANGES_REQUESTED** [review]

File: grpc/send_incoming_call_notification.go
Comment (hauntii) [reply grpc/send_incoming_call_notification.go]: I hadn't tested this, but in fact, it is! It's the amount of times the PID gets repeated in the application data. Don't know why.

This also resolves the other conversation.

File: grpc/send_incoming_call_notification.go
Comment (hauntii) [reply grpc/send_incoming_call_notification.go]: These flags are unknown to me (also brought up in friends 3ds).
If anyone *does* know them a flag system would probably be handy.

File: grpc/send_incoming_call_notification.go
Comment (hauntii) [reply grpc/send_incoming_call_notification.go]: Whoops, slipped my brain a lil. These are now all plogger excluding the protocol / method logging in `server.go`, as in every other server that continues to use fmt.

File: grpc/send_incoming_call_notification.go
Comment (jonbarrow) [inline grpc/send_incoming_call_notification.go]: I wonder if it's a coincidence that `0x65` is 101, which is the `RequestJoinGathering` notification category?

**Review by jonbarrow — APPROVED** [review]

File: grpc/send_incoming_call_notification.go
Comment (jonbarrow) [reply grpc/send_incoming_call_notification.go]: Looks like some flags _are_ known actually https://nintendo-wiki.pretendo.network/docs/nex/protocols/friends-wiiu#changed-flags

`0x1FF` seems to contain all flags in this list, which feels odd?

File: grpc/send_incoming_call_notification.go
Comment (hauntii) [reply grpc/send_incoming_call_notification.go]: Would make sense. I'm going to perform a test to see if sending `102` stops the ringing as well.

Comment (hauntii) [reply grpc/send_incoming_call_notification.go]: Appears it does not correlate with the notification type, 102 does not stop ringing.

Comment (hauntii) [reply grpc/send_incoming_call_notification.go]: Removing this *does* stop the ringing however, so I'm unsure if they're related.

File: grpc/send_incoming_call_notification.go
Comment (hauntii) [reply grpc/send_incoming_call_notification.go]: This actually has no effect here, so I've removed it.

**Pull request merged**
